### PR TITLE
Do not apply the canvas bounds as a cull rectangle when recording a Skia picture

### LINKED ISSUE #12618: Material sheet shadows not rendering properly in ListView
## Steps to Reproduce

1.  Run the app below.
2.  See that the [Material](https://docs.flutter.io/flutter/material/Material-class.html) sheet shadows render properly.
3.  Scroll and see that the shadows disappear.

I noticed that the shadows *do* properly render if you do one of the following:
- Comment out both the subtitle and trailing elements of the ListTile - having either one causes the issue, but with both gone the shadows render.
- Wrap each Sheet with bottom padding instead of inserting it between elements.

Also interesting, if you fling the list, it looks like the elements coming on screen at the edge of the list *do* have the shadow visible, but it quickly disappears when they get painted again.

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(new MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: 'Shadow bug reproduction',
      theme: new ThemeData(
        primarySwatch: Colors.blue,
        scaffoldBackgroundColor: Colors.grey.shade100,
      ),
      home: new MyHomePage(),
    );
  }
}

const padding = const Padding(padding: const EdgeInsets.only(top: 20.0));

class MyHomePage extends StatelessWidget {

  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      appBar: new AppBar(
        title: new Text('Shadow behavior repro'),
      ),
      body: new Column(
        children: [
          new Expanded(
            child: new ListView(
              children: new List.filled(20, new Sheet())
                  .expand((sheet) => [padding, sheet]).toList(),
            ),
          ),
        ],
      ),
    );
  }
}

class Sheet extends StatelessWidget {
  @override
  Widget build(BuildContext context) => new Material(
      elevation: 2.0,
      child: new ListTile(
        title: const Text('List Tile Title'),
        // Note commenting out both subtitle and trailing
        // 'fixes' the shadow issue - they render properly w/ scrolling.
        subtitle: const Text('List Tile Subtitle'),
        trailing: const ExpandIcon(),
      ),
    );
}
```

Before scrolling:
image

After scrolling:
image

### PR DESCRIPTION
**Pull request by jason-simmons**
The picture may need to draw outside the bounds (specifically for shadows)

Fixes https://github.com/flutter/flutter/issues/12618

### CODE REVIEW

**Comment by jason-simmons** [comment]
@Hixie

File: packages/flutter/lib/src/rendering/object.dart [hunk 1]
```diff
@@ -59,25 +59,19 @@ typedef void PaintingContextCallback(PaintingContext context, Offset offset);
...
+  /// will record painting commands.  This can be useful for debugging.
   ///
-  /// A render object provided with this [PaintingContext] (e.g. in its
-  /// [RenderObject.paint] method) is permitted to paint outside the region that
-  /// the render object occupies during layout, but is not permitted to paint
-  /// outside these canvas paints bounds. These paint bounds are used to
-  /// construct memory-efficient composited layers, which means attempting to
-  /// paint outside these bounds can attempt to write to pixels that do not
-  /// exist in the composited layer.
+  /// Note that the canvas will allow painting outside these bounds.
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/object.dart]: s/Note that the/The/

**Comment by Hixie** [comment]
We should also change the docs of RenderObject.paintBounds

**Comment by Hixie** [comment]
(and maybe change its name as well)

File: packages/flutter/lib/src/rendering/object.dart [hunk 2]
```diff
@@ -210,9 +202,9 @@ class PaintingContext {
 
   void _startRecording() {
     assert(!_isRecording);
-    _currentLayer = new PictureLayer(canvasBounds);
+    _currentLayer = new PictureLayer(estimatedBounds);
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/object.dart]: can we remove the bounds stuff from the layers entirely?

File: packages/flutter/lib/src/rendering/object.dart [hunk 3]
```diff
@@ -309,7 +301,7 @@ class PaintingContext {
     assert(painter != null);
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/object.dart]: childPaintBounds should also have its documentation updated

**Comment by jason-simmons** [comment]
Updated - PTAL

**Comment by Hixie** [comment]
LGTM, but travis is angry.

**Comment by Hixie** [comment]
failure was intermittent, restarted

File: packages/flutter/lib/src/rendering/object.dart [hunk 4]
```diff
@@ -210,9 +202,9 @@ class PaintingContext {
 
   void _startRecording() {
     assert(!_isRecording);
-    _currentLayer = new PictureLayer(canvasBounds);
+    _currentLayer = new PictureLayer(estimatedBounds);
     _recorder = new ui.PictureRecorder();
-    _canvas = new Canvas(_recorder, canvasBounds);
+    _canvas = new Canvas(_recorder, new Rect.largest());
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/object.dart]: with https://github.com/flutter/engine/pull/4551 we can just omit the second argument, which is probably simpler.

**Comment by jason-simmons** [comment]
Updated for https://github.com/flutter/engine/pull/4551

**Comment by Hixie** [comment]
LGTM

**Comment by Hixie** [comment]
This caused a ~20% regression in performance. This was expected (we're trading performance for correctness), and so I've rebaselined the benchmarks.

**Pull request merged**